crates/ruff_linter/src/rules/pylint/rules/binary_op_and_normal_assignment.rs

```diff
@@ -0,0 +1,230 @@
+use ast::{Expr, StmtAugAssign};
+use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for normal assignment statements whose target is the same as the
+/// left operand of a binary operator, in which cases, augmented assignment
+/// could potentially be used instead.
+///
+/// ## Why is this bad?
+/// Augmented assignment operators are more concise to perform a binary
+/// operation and assign results back to one of the operands.
+///
+/// ## Example
+/// ```python
+/// a = a + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a += 1
+/// ```
+///
+/// ## Fix safety
+/// This rule's fix is marked as being unsafe, in that it could alter semantics
+/// of the given Python code in some scenarios.
+///
+/// For example, the following code using mutable data types as the assignment
+/// target
+/// ```python
+/// a = [1]
+/// b = a
+/// a = a + [2]
+/// assert (a, b) == ([1, 2], [1])
+/// ```
+///
+/// is not the same as
+/// ```python
+/// a = [1]
+/// b = a
+/// a += [2]
+/// assert (a, b) == ([1, 2], [1, 2])
+/// ```
+#[violation]
+pub struct BinaryOpAndNormalAssignment;
+
+impl Violation for BinaryOpAndNormalAssignment {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Normal assignment with left operand of binary operator being the same as the target."
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use augmented assignment instead.".to_string())
+    }
+}
+
+pub(crate) fn binary_op_and_normal_assignment(
+    checker: &mut Checker,
+    assign @ ast::StmtAssign { value, targets, .. }: &ast::StmtAssign,
+) {
+    if targets.len() != 1 {
```

Comment:
Using [let-else](https://doc.rust-lang.org/beta/rust-by-example/flow_control/let_else.html) here to make this check a bit nicer:

```rust
let target = Some(targets.first()) else {
    return;
}
```

Reply:
I don't think they are semantically equal. In particular, the following test case which is marked as `OK` fails after the suggested change.

```python
a_number = index = a_number + 1 # OK
```

I think a further question would be whether the should detect the above case and, in the same gist, the one below.

```python
index = a_number = a_number + 1 # OK
```

Reply:
I had a quick test with `pylint 3.0.3`, whose `R6104` rule detects the 

```python
a_number = index = a_number + 1 # OK
```

but not this,
```python
index = a_number = a_number + 1 # OK
```

Do you think we should implement to reproduce this effect?